File: l4m-esports-angular/src/app/core/services/socket.service.ts
import { Injectable, inject } from '@angular/core';
import { io, Socket } from 'socket.io-client';
import { AuthService } from './auth.service';
import { BehaviorSubject, Observable } from 'rxjs';

@Injectable({ providedIn: 'root' })
export class SocketService {
  private socket: Socket | null = null;
  private connected = new BehaviorSubject<boolean>(false);
  private readonly authService = inject(AuthService);

  connect(): void {
    if (this.socket?.connected) {
      return;
    }

    const token = this.authService.token;
    this.socket = io('http://localhost:3000', {
      auth: { token },
      transports: ['websocket', 'polling']
    });

    this.socket.on('connect', () => {
      console.log('🔌 Socket.io connected', this.socket?.id);
      this.connected.next(true);
    });

    this.socket.on('disconnect', (reason) => {
      console.log('🔌 Socket.io disconnected', reason);
      this.connected.next(false);
    });

    this.socket.on('connect_error', (error) => {
      console.error('🔌 Socket.io connection error:', error);
    });
  }

  disconnect(): void {
    if (this.socket) {
      this.socket.disconnect();
      this.socket = null;
      this.connected.next(false);
    }
  }

  on(event: string): Observable<any> {
    return new Observable((observer) => {
      if (!this.socket) {
        this.connect();
      }

      const handler = (data: any) => {
        console.log(`🔔 Socket.io event received: ${event}`, data);
        observer.next(data);
      };

      this.socket!.on(event, handler);
      console.log(`👂 Listening to Socket.io event: ${event}`);

      return () => {
        if (this.socket) {
          this.socket.off(event, handler);
        }
      };
    });
  }

  isConnected(): Observable<boolean> {
    return this.connected.asObservable();
  }

  reconnect(): void {
    this.disconnect();
    this.connect();
  }
}

File: l4m-esports-mobile/app/src/main/java/com/example/l4m_esports_mobile/data/remote/SocketService.kt
package com.example.l4m_esports_mobile.data.remote

import android.util.Log
import com.example.l4m_esports_mobile.data.local.PreferencesManager
import com.example.l4m_esports_mobile.util.Constants
import io.socket.client.IO
import io.socket.client.Socket
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.runBlocking
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class SocketService @Inject constructor(
    private val preferencesManager: PreferencesManager
) {
    private var socket: Socket? = null
    private val _isConnected = MutableStateFlow(false)
    val isConnected: StateFlow<Boolean> = _isConnected.asStateFlow()

    fun connect() {
        if (socket?.connected() == true) {
            return
        }

        try {
            val token = runBlocking { preferencesManager.getToken() }
            val options = IO.Options().apply {
                auth = mapOf("token" to (token ?: ""))
                transports = arrayOf("websocket", "polling")
            }

            socket = IO.socket("${Constants.BASE_URL}", options)

            socket?.on(Socket.EVENT_CONNECT) {
                Log.d("SocketService", "🔌 Socket.io connected")
                _isConnected.value = true
            }

            socket?.on(Socket.EVENT_DISCONNECT) {
                Log.d("SocketService", "🔌 Socket.io disconnected")
                _isConnected.value = false
            }

            socket?.on(Socket.EVENT_CONNECT_ERROR) { args ->
                Log.e("SocketService", "🔌 Socket.io connection error: ${args[0]}")
            }

            socket?.connect()
        } catch (e: Exception) {
            Log.e("SocketService", "Error connecting to socket: ${e.message}")
        }
    }

    fun disconnect() {
        socket?.disconnect()
        socket = null
        _isConnected.value = false
    }

    fun on(event: String, callback: (Array<out Any>) -> Unit) {
        socket?.on(event) { args ->
            callback(args)
        }
    }

    fun off(event: String) {
        socket?.off(event)
    }

    fun reconnect() {
        disconnect()
        connect()
    }
}

